src/include/daos/checksum.h

```diff
@@ -15,48 +15,259 @@
+	CSUM_TYPE_UNKNOWN = 0,
+
+	CSUM_TYPE_ISAL_CRC16_T10DIF = 1,
+	CSUM_TYPE_ISAL_CRC32_ISCSI = 2,
+	CSUM_TYPE_ISAL_CRC64_REFL = 3,
+	CSUM_TYPE_ISAL_SHA1 = 4,
+
+	CSUM_TYPE_END = 5,
+};
 
-enum {
-	DAOS_CS_CRC32 = 0,
-	DAOS_CS_CRC64 = 1,
-	DAOS_CS_MAX,
-	DAOS_CS_UNKNOWN,
+
+/** Lookup the appropriate CSUM_TYPE given daos container property */
+enum DAOS_CSUM_TYPE daos_contprop2csumtype(int contprop_csum_val);
+
+struct csum_ft;
+struct daos_csummer {
+	/** Size of csum_buf. */
+	size_t dcs_csum_buf_size;
+	/** Cached configuration for chunk size*/
+	size_t dcs_chunk_size;
+	/** Pointer to the function table to be used for calculating csums */
+	struct csum_ft *dcs_algo;
+	/** Pointer to function table specific contexts */
+	void *dcs_ctx;
+	/** Points to the buffer where the  calculated csum is to be written */
+	uint8_t *dcs_csum_buf;
 };
 
-struct daos_csum {
-	int			dc_init:1;
-#if defined(__x86_64__)
-	int			dc_csum;
-#else
-	mchecksum_object_t	dc_csum;
-#endif
-	char			dc_buf[DAOS_CSUM_SIZE];
+struct csum_ft {
+	int (*init)(struct daos_csummer *obj);
+	void (*destroy)(struct daos_csummer *obj);
+	int (*finish)(struct daos_csummer *obj);
+	int (*update)(struct daos_csummer *obj, uint8_t *buf, size_t buf_len);
+	void (*reset)(struct daos_csummer *obj);
+	void (*get)(struct daos_csummer *obj);
+	size_t (*get_size)(struct daos_csummer *obj);
 
+	/** Len in bytes. Ft can either statically set csum_len or provide
+	 *  a get_len function
+	 */
+	size_t csum_len;
+	char *name;
+	uint16_t type;
 };
 
-typedef struct daos_csum daos_csum_t;
-int		daos_csum_init(const char *cs_name, daos_csum_t *checksum);
-int		daos_csum_free(daos_csum_t *csum);
-int		daos_csum_reset(daos_csum_t *csum);
-int		daos_csum_compute(daos_csum_t *csum, d_sg_list_t *sgl);
-daos_size_t	daos_csum_get_size(const daos_csum_t *csum);
-int		daos_csum_get(daos_csum_t *csum, daos_csum_buf_t *csum_buf);
-int		daos_csum_compare(daos_csum_t *csum, daos_csum_t *csum_src);
-#endif
+struct csum_ft *
+daos_csum_type2algo(enum DAOS_CSUM_TYPE type);
+
+/**
+ * -----------------------------------------------------------------------------
+ * daos_csummer Functions
+ * -----------------------------------------------------------------------------
+ */
+/**
+ * Initialize the daos_csummer
+ *
+ * @param obj		daos_csummer to be initialized. Memory will be allocated
+ *			for it.
+ * @param ft		Pointer to the function table for checksum calculations
+ * @param chunk_bytes	Chunksize, typically from the container configuration
+ *
+ * @return		0 for success, or an error code
+ */
+int
+daos_csummer_init(struct daos_csummer **obj, struct csum_ft *ft,
+		  size_t chunk_bytes);
+
+/** Destroy the daos_csummer */
+void
+daos_csummer_destroy(struct daos_csummer **obj);
+
+/** Get the checksum length from the configured csummer. */
+size_t
+daos_csummer_get_size(struct daos_csummer *obj);
+
+/** Determine if the checksums is configured. */
+bool
+daos_csummer_get_is_set(struct daos_csummer *obj);
```

Comment:
how about something like daos_csummer_initialized?

Reply:
Yes, that's way better. Will update.